File: LeetcodePractice/src/main/java/com/Bernie/leetcode/editor/cn/[54]螺旋矩阵.py
from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        ans = list()
        if not matrix or not matrix[0]:
            return ans
        row, columns = len(matrix), len(matrix[0])
        left, right, top, bottom = 0, columns - 1, 0, row - 1
        while left <= right and top <= bottom:
            # 从左到右
            for column in range(left, right + 1):
                ans.append(matrix[top][column])
            # 从上到下
            for row in range(top + 1, bottom + 1):
                ans.append(matrix[row][right])
            if left < right and top < bottom:
                # 从右向左
                for column in range(right - 1, left - 1, -1):
                    ans.append(matrix[bottom][column])
                # 从下往上
                for row in range(bottom - 1, top, -1):
                    ans.append(matrix[row][left])

            left, right, top, bottom = left + 1, right - 1, top + 1, bottom - 1

        return ans


solution = Solution()
ans = solution.spiralOrder([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(ans)

File: LeetcodePractice/src/main/java/com/Bernie/leetcode/editor/cn/[54]螺旋矩阵.java
package com.Bernie.leetcode.editor.cn;
//给你一个 m 行 n 列的矩阵 matrix ，请按照 顺时针螺旋顺序 ，返回矩阵中的所有元素。
//
// 
//
// 示例 1： 
//
// 
//输入：matrix = [[1,2,3],[4,5,6],[7,8,9]]
//输出：[1,2,3,6,9,8,7,4,5]
// 
//
// 示例 2： 
//
// 
//输入：matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
//输出：[1,2,3,4,8,12,11,10,9,5,6,7]
// 
//
// 
//
// 提示： 
//
// 
// m == matrix.length 
// n == matrix[i].length 
// 1 <= m, n <= 10 
// -100 <= matrix[i][j] <= 100 
// 
// Related Topics 数组 
// 👍 670 👎 0


import java.util.ArrayList;
import java.util.List;

//leetcode submit region begin(Prohibit modification and deletion)
class SpiralMatrix {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ans = new ArrayList<>();
        if (null == matrix || matrix.length == 0 || matrix[0].length == 0) {
            return ans;
        }
        int rows = matrix.length, columns = matrix[0].length;
        int left = 0, right = columns - 1, top = 0, bottom = rows -1;

        while (left <= right && top <= bottom) {
            // 从左向右遍历
            for (int i = left; i <= right; i++) {
                ans.add(matrix[top][i]);
            }
            // 从上到下遍历
            for (int i = top + 1; i <= bottom; i++) {
                ans.add(matrix[i][right]);
            }

            if (left < right && top < bottom) {
                // 从右向左遍历
                for (int i = right - 1; i > left; i--) {
                    ans.add(matrix[bottom][i]);
                }

                // 从下到上
                for (int i = bottom; i > top ; i--) {
                    ans.add(matrix[i][left]);
                }
            }

            left++;
            right--;
            top++;
            bottom--;

        }

        return ans;
    }
}
//leetcode submit region end(Prohibit modification and deletion)
